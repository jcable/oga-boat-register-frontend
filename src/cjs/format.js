const M2F = 3.2808;

const pounds = new Intl.NumberFormat('en-GB', { currency: 'GBP', style: 'currency' });

module.exports.price = function (n) {
    if (n === 0) {
        return 'offers';
    }
    return pounds.format(n);
}

module.exports.feet = function(n) {
    return `${n.toFixed(2)} ft`;
}

module.exports.m2df = function(val) {
    if(val) {
        return (M2F*val).toFixed(2);
    }
}

module.exports.m2dsqf = function(val) {
    if(val) {
        return (M2F*M2F*val).toFixed(2);
    }
}

module.exports.m2dfn = function(val) {
    if(val) {
        return parseFloat((M2F*val).toFixed(2));
    }
}

module.exports.m2dsqfn = function(val) {
    if(val) {
        return parseFloat((M2F*M2F*val).toFixed(3));
    }
}

module.exports.f2m = function(val) {
    if(val) {
        return Math.round(1000*val/M2F) / 1000;
    }
}

module.exports.f2m2 = function(val) {
    if(val) {
        return Math.round(1000*val/M2F/M2F) / 1000;
    }
}