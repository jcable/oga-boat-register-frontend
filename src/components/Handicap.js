import { componentTypes } from "@data-driven-forms/react-form-renderer";

const handicapForm = {
    name: "handicap",
    title: "Handicap Data",
    description: `This form collects the information needed to calculate a traditional TH(C)F
handicap and the extra data for experimental and area handicaps. You can enter the data in either decimal feet or in metres.`,
    component: componentTypes.SUB_FORM,
    fields: [
      {
        name: 'units',
        component: componentTypes.SWITCH,
        label: "Enter values in",
        onText: 'decimal feet',
        offText: 'metres'  
      },
      {
        component: componentTypes.RADIO,
        name: "ddf_collect_headsail_data",
        label: "Collect headsail data",
        "options": [
          {label: 'Yes', value: true},
          {label: 'No', value: false},
        ],
      }
    ]
  };

  const sailAreaForm = {
      title: "Sail Area",
      name: "sa",
      component: componentTypes.SUB_FORM,
      fields: [
        {
          component: componentTypes.TEXT_FIELD,
          name: "handicap_data~sailarea",
          label: "Area in square feet",
          dataType: 'float'
        },
        {
          component: componentTypes.TEXT_FIELD,
          name: "handicap_data~fore_triangle_height",
          label: "Fore Triangle Height",
          dataType: 'float'
        },
        {
          component: componentTypes.TEXT_FIELD,
          name: "handicap_data~fore_triangle_base",
          label: "Fore Triangle Base",
          dataType: 'float'
        },
      ]
    };
    /*
    length_overall
    length_on_waterline
    length_over_spars
    moving_keel
    draft_keel_up
    draft_keel_down
    "moving_keel_type": Centreboard|Leeboard, 
    calculated_thcf
  "main": {"foot": 15.9, "head": 11.3, "luff": 13.4}, 
mizzen": {"foot": 3.33, "luff": 7.5}
 sailarea
 "topsail": {"luff": 16, "perpendicular": 12}

    */

    const luffLeachFootForm = (title, name) => {
      return {
      title,
      name,
      component: componentTypes.SUB_FORM,
      fields: [
        {
          component: componentTypes.TEXT_FIELD,
          name: `${name}~luff`,
          label: "Luff",
          dataType: 'float'
        },
        {
          component: componentTypes.TEXT_FIELD,
          name: `${name}~leach`,
          label: "Leach",
          dataType: 'float'
        },
        {
          component: componentTypes.TEXT_FIELD,
          name: `${name}~foot`,
          label: "Foot",
          dataType: 'float'
        },
      ]
    };
  };

  const gaffForm = (title, name) => {
    return {    
      title,
      name,
      component: componentTypes.SUB_FORM,
      fields: [
        {
          component: componentTypes.TEXT_FIELD,
          name: `${name}~luff`,
          label: "Luff",
          dataType: 'float'
        },
        {
          component: componentTypes.TEXT_FIELD,
          name: `${name}~head`,
          label: "Head",
          dataType: 'float'
        },
        {
          component: componentTypes.TEXT_FIELD,
          name: `${name}~foot`,
          label: "Foot",
          dataType: 'float'
        },
      ]
    }
  };

  const topslForm = (title, name) => {
    return {
      title,
      name,
      component: componentTypes.SUB_FORM,
      fields: [
        {
          component: componentTypes.TEXT_FIELD,
          name: `${name}~luff`,
          label: "Luff",
          dataType: 'float'
        },
        {
          component: componentTypes.TEXT_FIELD,
          name: `${name}~perpendicular`,
          label: "Perpendicular",
          dataType: 'float'
        },
      ]
    }
  }

const propellorForm = {
      title: "Propeller type",
      name: "handicap_data~prop",
      component: componentTypes.SUB_FORM,
      fields: [
        {
          component: componentTypes.RADIO,
          name: "propeller_type",
          label: "Select",
          "options": [
            {label: 'None', value: 'none'},
            {label: 'Fixed', value: 'fixed'},
            {label: 'Folding', value: 'folding'},
            {label: 'Feathering', value: 'feathering'},
          ],
        }
      ]
    };
  
  export const steps = (pickers) => [
    {
      name: "handicap-step",
      nextStep: "hull_measurements-step",  
      nextStep: "sa-step",
      fields: [handicapForm],
    },
    {
      name: "sa-step",
      nextStep: "bigs-step",
      fields: [sailAreaForm],
    },
    {
      name: "bigs-step",
      nextStep: "bigj-step",
      fields: [luffLeachFootForm('Biggest Staysail', 'handicap_data~biggest_staysail')]
    },
    {
      name: "bigj-step",
      nextStep: "bigdw-step",
      fields: [luffLeachFootForm('Biggest Jib', 'handicap_data~biggest_jib')]
    },
    {
      name: "bigdw-step",
      nextStep: "main-step",
      fields: [luffLeachFootForm('Biggest Downwind sail', 'handicap_data~biggest_downwindsail')]
    },
    {
      name: "main-step",
      nextStep: "top-step",
      fields: [gaffForm('Main sail', 'handicap_data~main')]
    },
    {
      name: "top-step",
      nextStep: "mizen-step",
      fields: [topslForm('topsail', 'handicap_data~topsail')]
    },
    {
      name: "mizen-step",
      nextStep: "miztop-step",
      fields: [gaffForm('Mizen', 'handicap_data~mizen')]
    },
    {
      name: "miztop-step",
      nextStep: "prop-step",
      fields: [topslForm('Mizen topsail', 'handicap_data~mizen_topsail')]
    },
    {
      name: "prop-step",
      fields: [propellorForm]
    }
  ];
