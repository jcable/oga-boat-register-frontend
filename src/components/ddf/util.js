import { componentTypes } from "@data-driven-forms/react-form-renderer";
import { mapPicker } from "../Rig";

export const extendableItems = ({pickers, name, label}) => {
    return [
      {
        component: componentTypes.SELECT,
        name,
        label,
        isReadOnly: false,
        isSearchable: true,
        isClearable: true,
        options: mapPicker(pickers[name]),
      },
      {
        component: componentTypes.TEXT_FIELD,
        condition: {
          when: name,
          isEmpty: true,
        },
        name: `new_${name}`,
        label: `a ${label.toLowerCase()} not listed`,
        isRequired: false,
      },
    ];
  };
  
  export const builderItems = (pickers) => extendableItems({pickers, name: 'builder', label: 'Builder'})
  export const designerItems = (pickers) => extendableItems({pickers, name: 'designer', label: 'Designer'})
  export const designClassItems = (pickers) => extendableItems({pickers, name: 'design_class', label: 'Design Class'})
