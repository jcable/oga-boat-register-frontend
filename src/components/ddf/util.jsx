import { componentTypes } from "@data-driven-forms/react-form-renderer";

export const mapPicker = (m) => {
  return m.map((val) => {
    if (val.id) {
      return { label: val.name, value: val.id }
    }
    return { label: val.name, value: val.name }
  });
}

export const constructionItems = (pickers) => {
  return [
  {
    component: componentTypes.SELECT,
    name: "construction_material",
    label: "Construction material",
    isReadOnly: false,
    isSearchable: true,
    isClearable: true,
    options: mapPicker(pickers.construction_material),
  },
  {
    component: componentTypes.SELECT,
    name: "construction_method",
    label: "Construction method",
    isReadOnly: false,
    isSearchable: true,
    isClearable: true,
    options: mapPicker(pickers.construction_method),
  },
  {
    component: componentTypes.SELECT,
    name: "spar_material",
    label: "Spar material",
    isReadOnly: false,
    isSearchable: true,
    isClearable: true,
    options: mapPicker(pickers.spar_material),
  },
  {
    component: componentTypes.TEXT_FIELD,
    name: "construction_details",
    label: "Construction details",
  },
];
};

export const extendableItems = ({pickers, name, label}) => {
    return [
      {
        component: componentTypes.SELECT,
        name,
        label,
        isReadOnly: false,
        isSearchable: true,
        isClearable: true,
        options: mapPicker(pickers[name]),
      },
      {
        component: componentTypes.TEXT_FIELD,
        condition: {
          when: name,
          isEmpty: true,
        },
        name: `new_${name}`,
        label: `a ${label.toLowerCase()} not listed`,
        isRequired: false,
      },
    ];
  };
  
  export const builderItems = (pickers) => extendableItems({pickers, name: 'builder', label: 'Builder'})
  export const designerItems = (pickers) => extendableItems({pickers, name: 'designer', label: 'Designer'})
  export const designClassItems = (pickers) => extendableItems({pickers, name: 'design_class', label: 'Design Class'})
