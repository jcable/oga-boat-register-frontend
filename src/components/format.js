const pounds = new Intl.NumberFormat('en-GB', { currency: 'GBP', style: 'currency' });

const M2F = 3.28084;

export function price(n) {
    if (n === 0) {
        return 'offers';
    }
    return pounds.format(n);
}

export function feet(n) {
    return `${n.toFixed(2)} ft`;
}

export function m2df(val) {
    if(val) {
        return (M2F*val).toFixed(2);
    }
}

export function m2dsqf(val) {
    if(val) {
        return (M2F*M2F*val).toFixed(2);
    }
}

export function f2m(val) {
    if(val) {
        return val/M2F;
    }
}

export function f2m2(val) {
    if(val) {
        return val/M2F/M2F;
    }
}