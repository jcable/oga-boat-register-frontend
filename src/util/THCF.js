const rig_allowance = (r) => {
    switch(r) {
        case 'cutter': return 0.96;
        case 'yawl': return 0.94;
        case 'schooner': return 0.92;
        case 'ketch': return 0.90;
        default: return 1.0;
    };
};

export function fGaffSA(sail) {
    if(sail) {
        const b = Number(sail.foot);
        const h = Number(sail.luff);
        const g = Number(sail.head);
        const d = Math.sqrt(b*b+h*h);
        return 0.5*b*h+0.5*g*d;    
    }
    return 0;
}

export function fTopSA(sail) {
    if(sail) {
        const i = Number(sail.perpendicular);
        const h = Number(sail.luff);
        return 0.5*i*h;    
    }
    return 0;
}

export function fForeTriangle(data) {
    if(data) {
        const i = Number(data.fore_triangle_height);
        const j = Number(data.fore_triangle_base);
        return 0.5*i*j;    
    }
    return 0;
}

// The measured sail area is the sum of:-
// 0.5(b x h) + 0.5(g x d) for mainsail and mizzen, 
// where d = √(b^2+h^2 ), plus 0.5(i x h) for topsails, 
// plus O.425(i x j) for the foretriangle.
export function fMSA(data) {
    if(data) {
        const sa = 0
        + fGaffSA(data.main)
        + fGaffSA(data.mizen)
        + fTopSA(data.topsail)
        + fTopSA(data.mizen_topsail)
        + 0.85*fForeTriangle(data);
        return sa;
    }
    return 0;
}

export function fL(data) {
    if(data && data.length_over_all && data.length_on_waterline ) {
        return 0.5*(data.length_over_all+data.length_on_waterline);
    }
    return 0;
}

export function fB(boat) {
    if(boat) {
        return boat.beam;
    }
    return 0;
}

export function fSqrtS(boat) {
    if(boat && boat.handicap_data) {
        const SA = boat.handicap_data.sailarea || fMSA(boat.handicap_data);
        return rig_allowance(boat.rig_type.toLowerCase())*Math.sqrt(SA);
    }
    return 0;
}

/*
export function fD(data) {
    if(data && data.depth) {
        return 1.25*data.depth;
    }
    return 0;
}
*/

export function fBD(boat) {
    // John Scarlett
    // return fB(boat)*fD(boat.handicap_data);
    // old online register
    return 0.67*boat.beam*boat.beam;
}

export function fMR(boat) {
    if(boat) {
        const L = fL(boat.handicap_data);
        const sqrtS = fSqrtS(boat);
        const BD = fBD(boat);
        if ( BD>0) {   
            const x = 0.15*L*sqrtS/Math.sqrt(BD);
            const y = 0.2*(L+sqrtS);    
            return x + y;
        }
    }
    return 0;
}

/* 
From the MR, bonuses, expressed as percentages are deducted.
R = MR - Bonus.
These bonuses are:- 
    for a fixed propeller – 3% ; 
    folding prop – 1½%; 
    built pre 1930 – 2½%; (removed)
        pre 1914 - 5%; (removed)
        pre 1900 – 7% ; (removed)
    draft less than 66% of 0.16 LWL - 4%; 
    less than 80% - 2%. 
    Shoal draft bonus halved if centreboard or leeboards fitted.
*/
export function fPropellerBonus(R, data) {
    if(data) {
        switch(data.prop) {
            case 'fixed': return 0.03*R;
            case 'folding': return 0.015*R;
            default: return 0;
        }    
    }
    return 0;
}

export function fShoalBonus(R, boat) {
    if(boat && boat.handicap_data) {
        const lwl = boat.handicap_data.length_on_waterline;
        const draft = boat.draft;
        const nominal_draft = 0.16 * lwl - 0.04 * lwl;
        let bonus = 0;
        if(draft < 0.8*nominal_draft) bonus = 0.02*R;
        if(draft < 0.66*nominal_draft) bonus = 0.02*R;
        if(['centre-boarder', 'centreboard dinghy', 'leeboarder'].includes(boat.hull_form)){
            bonus = bonus / 2;
        }
        return bonus;    
    }
    return 0;
}

export function fR(boat) {
    if(boat) {
        const MR = fMR(boat);
        return MR
        - fPropellerBonus(MR, boat.handicap_data)
        ;//- fShoalBonus(MR, boat);    
    }
    return 0;
}

export function thcf(boat) {
    if(boat) {
        return 0.125*(Math.sqrt(fR(boat))+3);
    }
    return 0;    
}