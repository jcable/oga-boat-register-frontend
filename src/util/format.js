const M2F = 3.2808;

const pounds = new Intl.NumberFormat('en-GB', { currency: 'GBP', style: 'currency' });

export function price(n) {
    if (n === 0) {
        return 'offers';
    }
    return pounds.format(n);
}

export function m2df(val) {
    if(val) {
        return (M2F*val).toFixed(2);
    }
}

export function feet(n) {
  return `${n.toFixed(2)} ft`;
}

export function m2feet(val) { 
  if(val) {
      return `${m2df(val)} ft`
  }
}

export function m2dsqf(val) {
    if(val) {
        return (M2F*M2F*val).toFixed(3);
    }
}

export function m2dfn(val) {
    if(val) {
        return Math.round(1000*M2F*val) / 1000;
    }
}

export function m2dsqfn(val) {
    if(val) {
        return parseFloat((M2F*M2F*val).toFixed(3));
    }
}

export function f2m(val) {
    if(val) {
        return val/M2F;
    }
}

export function f2m2(val) {
    if(val) {
        return val/M2F/M2F;
    }
}


const metreKeys = [
    'beam','draft','perpendicular','luff','head','leech','foot',
    'length_on_deck','length_on_waterline','length_over_all',
    'fore_triangle_height','fore_triangle_base',
  ];
  const squareMetreKeys = ['sailarea'];
  const booleanKeys = ['year_is_approximate'];
  const keysToOmit = [
    'builderByBuilder','constructionMaterialByConstructionMaterial',
    'constructionMethodByConstructionMethod','designClassByDesignClass',
    'designerByDesigner','genericTypeByGenericType','rigTypeByRigType',
  ];
  /*
  biggest staysail (luff, leech, foot)
  biggest jib (luff, leech, foot)
  biggest downwind sail (luff, leech, foot)
  */
  export function boatm2f(obj) {
    if(obj) {
      if(Array.isArray(obj)) {
        return obj.map((n) => boatm2f(n))
      } else if (typeof obj === 'object') {
        const r = {};
        Object.keys(obj).forEach(k => {
          if(metreKeys.includes(k)) {
            r[k] = m2dfn(obj[k]);
          } else if(squareMetreKeys.includes(k)) {
            r[k] = m2dsqfn(obj[k]);
          } else if(booleanKeys.includes(k)) {
            r[k] = !!obj[k];
          } else if (obj[k]){
            r[k] = boatm2f(obj[k]);
          }
        });
        return r;
      } else {
        return obj;
      }
    } else {
      return undefined;
    }
  }
  
  export function boatf2m(obj) {
    if(obj) {
      if(Array.isArray(obj)) {
        return obj.map((n) => boatm2f(n))
      } else if (typeof obj === 'object') {
        const r = {};
        Object.keys(obj).forEach(k => {
          if(metreKeys.includes(k)) {
            r[k] = f2m(obj[k]);
          } else if(squareMetreKeys.includes(k)) {
            r[k] = f2m2(obj[k]);
          } else if(booleanKeys.includes(k)) {
            r[k] = !!obj[k];
          } else {
            if (obj[k]) {
              r[k] = boatf2m(obj[k]);
            }
          } 
        });
        return r;
      } else {
        return obj;
      }
    }
    return obj;
  }
  
  export function boatDefined(obj) {
    if(obj) {
      if(Array.isArray(obj)) {
        return obj.map((n) => boatDefined(n))
      } else if (typeof obj === 'object') {
        const r = {};
        Object.keys(obj).forEach(k => {
          if(metreKeys.includes(k)) {
            r[k] = obj[k];
          } else if(squareMetreKeys.includes(k)) {
            r[k] = obj[k];
          } else if(booleanKeys.includes(k)) {
            r[k] = !!obj[k];
          } else if (keysToOmit.includes(k)) {
            // console.log('omitting', k);
          } else {
            if (obj[k]) {
              r[k] = boatDefined(obj[k]);
            }
          } 
        });
        return r;
      } else {
        return obj;
      }
    }
    return obj;
  }
  
  export function m2fall(o) {
    if(o) {
        return Object.keys(o).map(k => m2dfn(o[k]));
    }
  }
