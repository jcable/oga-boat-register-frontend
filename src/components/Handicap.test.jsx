import { foretriangle_area } from './Handicap';
import { thcf, fGaffSA, fTopSA, fForeTriangle, fMSA, fL, fMR, fBD, fSqrtS, fR } from '../util/THCF';
import { boatm2f, boatf2m } from "../util/format";

const Transcur = { "id": "656e0e94-c298-408e-90e2-1bafd5bdd824", "name": "Transcur", "year": 1889, "year_is_approximate": false, "place_built": "Brightlingsea", "home_port": "Pin Mill", "home_country": "GBR", "sail_number": "CK365", "oga_no": 2932, "fishing_number": "CK365", "full_description": "<p>Transcur CK 365 was built in 1889, it seems to replace a larger 17 ton smack of the same name that was registered out of Maldon as MN11 in 1871. It was in 1898 that the Maldon &#39;Transcur&#39; did not appear in the register and &#39;Transcur CK 365&#39; was first registered.</p>\r\n\r\n<p>Her early history is rather vague and the only thing we know is that she worked from the Essex River Crouch and was owned by the Smith Brothers, well known oyster merchants of the area. In 1924 she came out of the register and was sold for conversion to a yacht. Coamings and skylights were added. In 1937 her first engine was fitted on the starboard quarter. She passed through several hands, each owner keeping her for approximatly five years. Most of this time was spent on the east coast of England.</p>\r\n\r\n<p>In 1998 the Thomas family bought Transcur from Lymington on the Solent where she had been for 19 years. Not in a sound enough condition to be sailed back to the east coast, she came back overland to the front garden where the family spent the next two years restoring her, keeping the floors but the rest of the boat was so rotten it needed replacing. She was returned to her original gaff rig which she lost in the late 50s. Since her restoration she has been sailed regularly to Holland each year for family holidays.</p>\r\n\r\n<p>Unlike many smacks, she is sailed by a family and cruised extensivly.</p>", "image_key": "grRdbV", "spar_material": "wood", "rig_type": "Cutter", "construction_material": "wood", "construction_method": "carvel", "construction_details": "Trad ", "draft": 1.396, "generic_type": "Smack", "handicap_data": { "propellor": { "type": "folding" }, "beam": 3.28, "main": { "foot": 7.01, "head": 5.029, "luff": 5.486 }, "thcf": 1.025, "draft": 1.402, "topsail": { "luff": 6.706, "perpendicular": 2.438 }, "sailarea": 81.065, "length_on_deck": 10.79, "fore_triangle_base": 7.315, "length_on_waterline": 8.687, "fore_triangle_height": 10.059 }, "hull_form": "long keel deep forefoot", "length_on_deck": 10.796, "mainsail_type": "gaff", "short_description": "'Transcur' was built in 1889 as an oyster dredger working on the east coast of England. She is just over 35ft long weighing 12 tons. 'Transcur' was totally rebuilt between 1999 and 2001, when she was relaunched.", "updated_at": "2021-02-07T07:39:42.886262+00:00", "beam": 3.301, "builder": "78bc9e97-d57d-481e-b855-b10261a92e4d", "designer": "74db1b7f-49fa-49d4-8e3f-9f172cd58673", "constructionMaterialByConstructionMaterial": { "name": "wood" }, "constructionMethodByConstructionMethod": { "name": "carvel" }, "designerByDesigner": { "name": "Aldous" }, "rigTypeByRigType": { "name": "Cutter" }, "genericTypeByGenericType": { "name": "Smack" }, "builderByBuilder": { "name": "Aldous Ltd." }, "for_sale_state": { "text": "not_for_sale" }, "for_sales": [], "engine_installations": [] };
const Robinetta = { "id": "85315666-4359-47f7-8d8d-4dd2905b7aa2", "name": "Robinetta", "year": 1937, "year_is_approximate": false, "place_built": "Rock Ferry Birkenhead", "home_port": "Tollesbury", "home_country": "GBR", "ssr": "131134", "sail_number": "315", "nsbr": "1717", "oga_no": 315, "callsign": "MKDZ8", "full_description": "\"When his last boat, the 42' Tredwin barge Pearl, was burnt out on moorings, Mr. Denys Rayner decided to try his hand at designing a boat, after studying the &quot;How To&quot; articles in the Y.M. Economy limited the size to less than 23&#39; overall, while a plea from the Shipmate was more or less an order for 6ft headroom.&quot;</p><p>The preceding extract from the February 1937 issue of Yachting Monthly was penned by Maurice Griffith.</p><p>Robinetta was designed from the outset for single handing, with all lines except the jib halyard let to the cockpit. She is strongly built with pine planking on an oak frame with an iron keel and a good deal of lead ballast. Her main cabin is cavernous for such a tiny boat. Part of this stems from the extension of the hull upwards to form the cabin sides. Griffith goes on to say &quot;Perhaps the most striking feature of this experimental design is the fullness of the hull, the pronounced tumblehome of the topsides from amidships aft, and the enormous quarters, which appear to be based on those of the French crabbers.&quot;</p><p>Robinetta is mostly original. Significant changes were made by Mike Garnham who replaced the previous engine by the current Yanmar 1GM10 and re-worked the cockpit. Major structural repairs were performed during the winter of 2007/2008 strengthening the keel fixings and doubling several ribs.</p><p>She was re-caulked with new gaff and sails in 2012.<br />Previous sail dimensions: Mainsail Head: 9.17 ft Topsail Perpendicular: 6ft Topsail luff 13 ft</p><p>Deck beam replaced and other repairs winter 2017/18.</p><p>Read articles about Robinetta from the Royal Cruising Club Journal, 1938, on <a href=\"http://www.sailing-by.org.uk/content/standing-waves-and-submarines\">Sailing by</a> and explore her <a href=\"http://robinetta-log.blogspot.co.uk/\" target=\"_blank\">blog</a> with a detailed history of ownership and other exploits.", "image_key": "9NV8KW", "uk_part1": "165315", "rig_type": "Cutter", "construction_material": "wood", "construction_method": "carvel", "spar_material": "wood", "construction_details": "Carvel pitch pine on oak and elm", "draft": 1.372, "generic_type": "Yacht", "handicap_data": { "propellor": { "type": "fixed" }, "beam": 2.438, "main": { "foot": 3.999, "head": 4.267, "luff": 4.319 }, "draft": 1.372, "length_on_deck": 6.858, "fore_triangle_base": 1.524, "length_on_waterline": 6.096, "fore_triangle_height": 4.572 }, "hull_form": "long keel deep forefoot", "length_on_deck": 6.858, "mainsail_type": "gaff", "sail_type": { "name": "gaff" }, "short_description": "<i>Robinetta</i> is a 22' 6\" tabloid cruiser, designed by Denys Rayner and built for him by the Enterprise Small Craft Company, Rock Ferry, Birkenhead.", "updated_at": "2021-01-31T17:02:41.408611+00:00", "website": "https://robinetta-log.blogspot.com/p/robinetta.html", "beam": 2.438, "builder": "a27bca50-d155-428a-be49-a542d64ba316", "designer": "afc68d59-e7ba-466e-861f-741e2eeae526", "constructionMaterialByConstructionMaterial": { "name": "wood" }, "constructionMethodByConstructionMethod": { "name": "carvel" }, "designerByDesigner": { "name": "D.A. Rayner" }, "rigTypeByRigType": { "name": "Cutter" }, "genericTypeByGenericType": { "name": "Yacht" }, "builderByBuilder": { "name": "Enterprise Small Craft" }, "for_sale_state": { "text": "not_for_sale" }, "for_sales": [], "engine_installations": [] };
/*
Luff 5.7 m or 18.696
Head 3 m or 9.84 feet
Leach 5.7 m or 29.2576 feet
Foot 5.24 m or 17.1872 feet
*/

test('boatm2f', () => {
  const a = { "1": 1, q: { a: 2 } };
  expect(boatm2f(a)).toStrictEqual(a);
});

const tcf = {
  "beam": 10.76,
  "draft": 4.6,
  "fore_triangle_base": 23.999,
  "fore_triangle_height": 33,
  "length_on_waterline": 28.5,
  "length_on_deck": 35.4,
  "propellor": { "type": "folding" },
  "main": {
    "foot": 22.998,
    "head": 16.499,
    "luff": 17.998,
  },
  "sailarea": 872.555,
  "thcf": 1.025,
  "topsail": {
    "luff": 22,
    "perpendicular": 8,
  },
};

const rcf = {
  "beam": 7.999,
  "draft": 4.501,
  "fore_triangle_base": 5,
  "fore_triangle_height": 15,
  "length_on_waterline": 20,
  "length_on_deck": 22.5,
  "propellor": { "type": "fixed" },
  "main": {
    "foot": 13.12,
    "head": 13.999,
    "luff": 14.17,
  },
};

test('boatm2f Transcur feet', () => {
  expect(boatm2f(Transcur.handicap_data)).toStrictEqual(tcf);
});

test('boatm2f Robinetta feet', () => {
  const m = Robinetta.handicap_data;
  const f = boatm2f(m);
  expect(f).toStrictEqual(rcf);
});

test('boatm2f Transcur metres', () => {
  const t = boatm2f(Transcur);
  expect(boatf2m(t)).toStrictEqual(Transcur);
});

test('boatm2f Robinetta metres', () => {
  expect(boatf2m(boatm2f(Robinetta))).toStrictEqual(Robinetta);
});

test('Transcur foretriangle_area', () => {
  const f = foretriangle_area(tcf);
  expect(f).toEqual(336.6);
});

test('Robinetta foretriangle_area', () => {
  const f = foretriangle_area(rcf);
  expect(f).toEqual(31.875);
});

test('Transcur fGaffSA', () => {
  expect(fGaffSA(tcf.main)).toEqual(447.95085079741887);
});
test('Transcur fTopSA', () => {
  expect(fTopSA(tcf.topsail)).toEqual(88);
});

test('Transcur fForeTriangle', () => {
  expect(Math.round(fForeTriangle(tcf))).toEqual(Math.round(336.6 / .85));
});
test('Transcur fMSA', () => {
  expect(Math.round(100 * fMSA(tcf))).toEqual(Math.round((100 * tcf.sailarea)));
});

test('Transcur fL', () => {
  expect(fL(tcf)).toEqual((tcf.length_on_waterline + tcf.length_on_deck) / 2);
});

test('Transcur fBD', () => {
  const b = boatm2f(Transcur);
  expect(fBD(b)).toEqual(.67 * b.handicap_data.beam * b.handicap_data.beam);
});

test('Transcur fSqrtS', () => {
  const x = .96 * Math.sqrt(tcf.sailarea);
  expect(fSqrtS(boatm2f(Transcur))).toEqual(x);
});

test('Transcur C', () => {
  expect(.67 * tcf.beam * tcf.beam).toEqual(77.570992);
});

test('Transcur fMR', () => {
  expect(fMR(boatm2f(Transcur))).toEqual(27.491938750055603);
});

test('Transcur fR', () => {
  expect(fR(boatm2f(Transcur))).toEqual(27.07955966880477);
});

test('Transcur T(H)CF', () => {
  expect(thcf(boatm2f(Transcur)).toFixed(3)).toEqual('1.025');
});

test('t1', () => {
  expect(0.125 * (Math.sqrt(27.04) + 3)).toEqual(1.025);
});

test('Robinetta T(H)CF', () => {
  expect(thcf(boatm2f(Robinetta)).toFixed(3)).toEqual('0.850');
});
