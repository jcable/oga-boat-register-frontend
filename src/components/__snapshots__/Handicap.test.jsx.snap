// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Robinetta T(H)CF 1`] = `"0.850"`;

exports[`Robinetta foretriangle_area 1`] = `2.961`;

exports[`Transcur C 1`] = `"NaN"`;

exports[`Transcur T(H)CF 1`] = `"1.025"`;

exports[`Transcur fForeTriangle 1`] = `"31.09"`;

exports[`Transcur fGaffSA 1`] = `"41.44"`;

exports[`Transcur fMR 1`] = `"27.49"`;

exports[`Transcur fMSA 1`] = `"76.0"`;

exports[`Transcur fR 1`] = `"27.08"`;

exports[`Transcur fSqrtS 1`] = `"28.35"`;

exports[`Transcur fTopSA 1`] = `"8"`;

exports[`Transcur foretriangle_area 1`] = `"31.1"`;

exports[`boatm2f Robinetta feet 1`] = `
Object {
  "beam": 8.009,
  "draft": 4.501,
  "fore_triangle_base": 5,
  "fore_triangle_height": 15,
  "length_on_deck": 22.5,
  "length_on_waterline": 20,
  "length_overall": 22.5,
  "main": Object {
    "foot": 13.123,
    "head": 13.999,
    "luff": 14.17,
  },
  "propellor": Object {
    "type": "fixed",
  },
  "sailarea": 260.013,
  "thcf": 0.85,
}
`;

exports[`boatm2f Transcur feet 1`] = `
Object {
  "beam": 10.761,
  "draft": 4.58,
  "fore_triangle_base": 23.999,
  "fore_triangle_height": 32.808,
  "length_on_deck": 35.42,
  "length_on_waterline": 28.501,
  "length_overall": 35.4,
  "main": Object {
    "foot": 22.966,
    "head": 16.404,
    "luff": 17.999,
  },
  "propellor": Object {
    "type": "folding",
  },
  "sailarea": 871.877,
  "thcf": 1.025,
  "topsail": Object {
    "luff": 22.001,
    "perpendicular": 7.999,
  },
}
`;

exports[`t1 1`] = `1.025`;
